Android/app/src/main/java/app/intra/socks/OverrideSocksHandler.java

```diff
@@ -0,0 +1,366 @@
+
+  void setContext(Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public void doConnect(Session session, CommandMessage commandMessage) throws IOException {
+    ServerReply reply = null;
+    Socket socket = null;
+    InetAddress bindAddress = null;
+    int bindPort = 0;
+    InetAddress remoteServerAddress = commandMessage.getInetAddress();
+    int remoteServerPort = commandMessage.getPort();
+
+    // set default bind address.
+    byte[] defaultAddress = {0, 0, 0, 0};
+    bindAddress = InetAddress.getByAddress(defaultAddress);
+    // DO connect
+    try {
+      // Connect directly.
+      socket = new Socket(remoteServerAddress, remoteServerPort);
+      bindAddress = socket.getLocalAddress();
+      bindPort = socket.getLocalPort();
+      reply = ServerReply.SUCCEEDED;
+
+    } catch (IOException e) {
+      if (e.getMessage().equals("Connection refused")) {
+        reply = ServerReply.CONNECTION_REFUSED;
+      } else if (e.getMessage().equals("Operation timed out")) {
+        reply = ServerReply.TTL_EXPIRED;
+      } else if (e.getMessage().equals("Network is unreachable")) {
+        reply = ServerReply.NETWORK_UNREACHABLE;
+      } else if (e.getMessage().equals("Connection timed out")) {
+        reply = ServerReply.TTL_EXPIRED;
+      } else {
+        reply = ServerReply.GENERAL_SOCKS_SERVER_FAILURE;
+      }
+      logger.info("SESSION[{}] connect {} [{}] exception:{}", session.getId(), new
+          InetSocketAddress(remoteServerAddress, remoteServerPort), reply, e.getMessage());
+    }
+
+    CommandResponseMessage responseMessage =
+        new CommandResponseMessage(VERSION, reply, bindAddress, bindPort);
+    session.write(responseMessage);
+    if (reply != ServerReply.SUCCEEDED) {
+      session.close();
+      return;
+    }
+
+    // Code prior to this point is essentially unmodified from sockslib.  At this point, sockslib
+    // normally creates a SocketPipe to pass the data.  We change the behavior here to allow retry.
+
+    final OutputStream sessionDownload = session.getOutputStream();
+    final InputStream sessionUpload = session.getInputStream();
+    InputStream socketDownload = socket.getInputStream();
+    OutputStream socketUpload = socket.getOutputStream();
+
+    final StreamPipe upload = new StreamPipe(sessionUpload, socketUpload, UPLOAD);
+    upload.setBufferSize(BUFFER_SIZE);
+    StreamPipe download = new StreamPipe(socketDownload, sessionDownload, DOWNLOAD);
+    download.setBufferSize(BUFFER_SIZE);
+
+    StatsListener listener = null;
+    try {
+      listener = runPipes(upload, download, remoteServerPort, SIMULATE_RESET && remoteServerPort == HTTPS_PORT);
+      if (remoteServerPort == HTTPS_PORT && listener.downloadBytes == 0 && listener.uploadBytes > 0
+          && listener.error != null && RESET_MESSAGE.equals(listener.error.getMessage())
+          && listener.uploadBuffer != null) {
+        socket.close();
```

Comment:
Very crafty.  It looks to me like there's not an easy way with sockslib to split the first upload packet before it gets sent on the original proxy-destination connection, and that's why you have to open a new proxy-destination connection on which to send the split packets?  If that's the correct reasoning, I think it would be helpful to put in a comment somewhere.

Reply:
That's not quite right.  My initial version split the first packet before it was sent, so it is possible.  However, this architecture allows us to apply the splitting only on sockets that have encountered the specific failure type.  This seems likely to have lower overhead than applying the splitting to all sockets.